### PR TITLE
Log Deprecations when Legacy API is used.

### PR DESCRIPTION
**Pull request by k2b6s9j**

### CODE REVIEW

**Comment by k2b6s9j** [comment]
I'm not sure which location for the deprecation works better. But they both seems to do it.

**Comment by k2b6s9j** [comment]
I tried to deprecate the public API methods themselves, but that placed the deprecations on Linter and the actual linter provider.

File: lib/linter.coffee [hunk 1]
```diff
@@ -2,6 +2,7 @@ fs = require 'fs'
 path = require 'path'
 {CompositeDisposable, Range, Point, BufferedProcess} = require 'atom'
 XRegExp = null
+Grim = require('grim')
```
Comment (steelbrain) [inline lib/linter.coffee]: `{deprecate} = require('grim')`

File: package.json [hunk 2]
```diff
@@ -32,6 +32,7 @@
     }
   },
   "dependencies": {
+    "grim": "^1.4.1",
```
Comment (steelbrain) [inline package.json]: You don't need to add it as a dep, It's available by default by atom I think.

File: lib/linter.coffee (hunk 1 above)
Comment (k2b6s9j) [inline lib/linter.coffee]: How come?

File: package.json (hunk 2 above)
Comment (k2b6s9j) [inline package.json]: Yeah... I thought the same, but it threw errors without the dep.

File: lib/linter.coffee (hunk 1 above)
Comment (steelbrain) [inline lib/linter.coffee]: I meant that you can use destructuring, we don't need that Grim variable for any other purposes than deprecate. Then how about you extract that variable from it on require time

Comment (k2b6s9j) [inline lib/linter.coffee]: Ah. I see. I'll do that.

**Comment by k2b6s9j** [comment]
Should be good to merge.

**Comment by k2b6s9j** [comment]
And get ready for the issues to come streaming in.

**Comment by steelbrain** [comment]
LGTM. Thoughts @AsaAyers

**Comment by k2b6s9j** [comment]
I guess we should cc @AtomLinter/linter

**Comment by AsaAyers** [comment]
I don't think this should be done until we have a complete (and verified) set of replacement tools for the CLI linters.

**Comment by AsaAyers** [comment]
If I were maintaining a CLI linter, I'd say we aren't ready to migrate

**Comment by k2b6s9j** [comment]
I'm maintaining a couple CLI Linters and they're working fine on the new API.

**Comment by steelbrain** [comment]
Yep, if only one of us had push access of `atom-linter` npm modle

**Comment by steelbrain** [comment]
@k2b6s9j It's true that most of the CLI linters work great, atom-hack is a CLI linter too, but some linters need methods like findFile that can only be implemented in an npm module.

**Comment by AsaAyers** [comment]
It's not the new API. it's the fact that you're copying and pasting significant portions of code to manage processing and parsing.

**Comment by k2b6s9j** [comment]
Ah... I see. We need to cut down the boilerplate needed for a CLI Linter.

**Comment by k2b6s9j** [comment]
I have an idea. Let me implement it.

**Comment by k2b6s9j** [comment]
Is it okay that my idea is service based rather than module based?

**Comment by steelbrain** [comment]
:sweat_smile: I am personally against that.

**Comment by AsaAyers** [comment]
As long as it gets implemented, I don't care at all how it's delivered. We have spent more time already fighting over whether it's a service or npm module than actually working on building it

**Comment by k2b6s9j** [comment]
I'll put up a basic WIP PR for my idea later tonight.

**Comment by k2b6s9j** [comment]
After dinner.

**Comment by park9140** [comment]
:ship:

**Comment by steelbrain** [comment]
Merging 'cause the next release will contain CLI helpers from #686

**Pull request merged**